File: backend/go/internal/mock/greeks.go
package mock

import "math"

// Greeks holds the first- and second-order Black-Scholes-Merton sensitivities.
// Conventions match src/services/greeksService.ts: vega per 1% vol; theta,
// charm, color per calendar day.
type Greeks struct {
	Price float64
	Delta float64
	Gamma float64
	Theta float64
	Vega  float64
	Rho   float64
	Charm float64
	Vanna float64
	Volga float64
	Speed float64
	Zomma float64
	Color float64
}

func normCDF(x float64) float64 { return 0.5 * (1 + math.Erf(x/math.Sqrt2)) }
func normPDF(x float64) float64 { return math.Exp(-0.5*x*x) / math.Sqrt(2*math.Pi) }

// BSM computes the option price and full Greek surface (dividend-free, q = 0).
func BSM(s, k, t, r, sigma float64, optType string) Greeks {
	isCall := optType == "CALL"
	if t <= 0 || sigma <= 0 {
		intrinsic := math.Max(0, s-k)
		if !isCall {
			intrinsic = math.Max(0, k-s)
		}
		return Greeks{Price: intrinsic}
	}
	const year = 365.0
	sqrtT := math.Sqrt(t)
	d1 := (math.Log(s/k) + (r+0.5*sigma*sigma)*t) / (sigma * sqrtT)
	d2 := d1 - sigma*sqrtT
	disc := math.Exp(-r * t)
	phi := normPDF(d1)

	var delta, price, thetaYr, rho float64
	if isCall {
		delta = normCDF(d1)
		price = s*normCDF(d1) - k*disc*normCDF(d2)
		thetaYr = -(s*phi*sigma)/(2*sqrtT) - r*k*disc*normCDF(d2)
		rho = k * t * disc * normCDF(d2) / 100
	} else {
		delta = normCDF(d1) - 1
		price = k*disc*normCDF(-d2) - s*normCDF(-d1)
		thetaYr = -(s*phi*sigma)/(2*sqrtT) + r*k*disc*normCDF(-d2)
		rho = -k * t * disc * normCDF(-d2) / 100
	}

	gamma := phi / (s * sigma * sqrtT)
	vega := s * phi * sqrtT / 100
	charmYr := -phi * ((2*r*t - d2*sigma*sqrtT) / (2 * t * sigma * sqrtT))
	colorYr := -phi / (2 * s * t * sigma * sqrtT) *
		(2*r*t + 1 + (d1*(2*r*t-d2*sigma*sqrtT))/(sigma*sqrtT))

	return Greeks{
		Price: price, Delta: delta, Gamma: gamma, Theta: thetaYr / year, Vega: vega, Rho: rho,
		Charm: charmYr / year, Vanna: -phi * (d2 / sigma), Volga: vega * (d1 * d2 / sigma),
		Speed: -(gamma / s) * (d1/(sigma*sqrtT) + 1), Zomma: gamma * ((d1*d2 - 1) / sigma), Color: colorYr / year,
	}
}

File: backend/python/common/greeks.py
"""Black-Scholes-Merton first- and second-order Greeks (dividend-free, q = 0).

Conventions mirror src/services/greeksService.ts: vega per 1% vol; theta,
charm, color per calendar day. Used by the options-chain service and the
backtesting engine so client and server agree to the last digit.
"""

from __future__ import annotations

import math
from dataclasses import asdict, dataclass

_YEAR = 365.0


def _norm_cdf(x: float) -> float:
    return 0.5 * (1.0 + math.erf(x / math.sqrt(2.0)))


def _norm_pdf(x: float) -> float:
    return math.exp(-0.5 * x * x) / math.sqrt(2.0 * math.pi)


@dataclass
class Greeks:
    price: float
    delta: float
    gamma: float
    theta: float
    vega: float
    rho: float
    lambda_: float
    epsilon: float
    charm: float
    vanna: float
    volga: float
    speed: float
    zomma: float
    color: float

    def as_dict(self) -> dict:
        d = asdict(self)
        d["lambda"] = d.pop("lambda_")  # reserved word in client schema
        return d


def bsm(spot: float, strike: float, t_years: float, rate: float, sigma: float, opt_type: str) -> Greeks:
    is_call = opt_type.upper() == "CALL"
    if t_years <= 0 or sigma <= 0:
        intrinsic = max(spot - strike, 0.0) if is_call else max(strike - spot, 0.0)
        return Greeks(intrinsic, 1.0 if (is_call and spot > strike) else 0.0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0)

    sqrt_t = math.sqrt(t_years)
    d1 = (math.log(spot / strike) + (rate + 0.5 * sigma * sigma) * t_years) / (sigma * sqrt_t)
    d2 = d1 - sigma * sqrt_t
    disc = math.exp(-rate * t_years)
    phi = _norm_pdf(d1)

    if is_call:
        delta = _norm_cdf(d1)
        price = spot * _norm_cdf(d1) - strike * disc * _norm_cdf(d2)
        theta_yr = -(spot * phi * sigma) / (2 * sqrt_t) - rate * strike * disc * _norm_cdf(d2)
        rho = strike * t_years * disc * _norm_cdf(d2) / 100.0
        epsilon = -spot * t_years * _norm_cdf(d1)
    else:
        delta = _norm_cdf(d1) - 1.0
        price = strike * disc * _norm_cdf(-d2) - spot * _norm_cdf(-d1)
        theta_yr = -(spot * phi * sigma) / (2 * sqrt_t) + rate * strike * disc * _norm_cdf(-d2)
        rho = -strike * t_years * disc * _norm_cdf(-d2) / 100.0
        epsilon = spot * t_years * _norm_cdf(-d1)

    gamma = phi / (spot * sigma * sqrt_t)
    vega = spot * phi * sqrt_t / 100.0
    lambda_ = delta * (spot / price) if price else 0.0
    charm_yr = -phi * ((2 * rate * t_years - d2 * sigma * sqrt_t) / (2 * t_years * sigma * sqrt_t))
    color_yr = (
        -phi
        / (2 * spot * t_years * sigma * sqrt_t)
        * (2 * rate * t_years + 1 + (d1 * (2 * rate * t_years - d2 * sigma * sqrt_t)) / (sigma * sqrt_t))
    )

    return Greeks(
        price=price,
        delta=delta,
        gamma=gamma,
        theta=theta_yr / _YEAR,
        vega=vega,
        rho=rho,
        lambda_=lambda_,
        epsilon=epsilon,
        charm=charm_yr / _YEAR,
        vanna=-phi * (d2 / sigma),
        volga=vega * (d1 * d2 / sigma),
        speed=-(gamma / spot) * (d1 / (sigma * sqrt_t) + 1),
        zomma=gamma * ((d1 * d2 - 1) / sigma),
        color=color_yr / _YEAR,
    )
